README.md
epi.branch.sim
==============

Simulates an epidemic outbreak with a stochastic branching process from
a number of initial seed cases.

### Features

Every simulated case has disease and intervention parameters
stochastically drawn from user-provided distributions. These parameters
determine which secondary cases occur, which in turn create the next
generation of cases. This simulation implements modelling of
interventions such as: \* manual/conventional contact tracing, \*
app-based contact tracing usage and compliance, \* reduction of contacts
through physical distancing, \* self-isolation of cases after certain
conditions are met, and \* quarantine/isolation of new arrivals.

This simulation also includes the option of adding new cases via
importation, which may have different contact tracing and self-isolation
conditions. The simulation itself outputs a dataframe that summarizes
every case generated during the course of the simulation. This can be
used to generate metrics or other statistics across a large number of
runs. Some example driver functions to do this are provided in the
package and as vignettes.

This package is inspired by [Hellewell et al. “Feasibility of
controlling COVID-19 outbreaks by isolation of cases and contacts.” *The
Lancet Global Health* 2020; 8: E488–E496. DOI:
10.1016/S2214-109X(20)30074-7](https://doi.org/10.1016/S2214-109X(20)30074-7).

### Installation

To install this package along with the vignettes, use

``` r
remotes::install_github("bcgov/epi.branch.sim", build_vignettes=TRUE)
```

### Usage

A very simple example of setting up and running a scenario for 30 days.

``` r
# Set up simulation objects
sim_params <- epi.branch.sim::initialize_sim_params(
  R0=3.0, infect_dur=999, vary_trace=FALSE, p_trace=0.8, 
  p_trace_app=0, p_trace_app_comp=0, p_symp=0.9, dt=1,
  incub_params=list(dist='weibull', shape=2.322737, scale=6.492272),
  generation_int_params=list(dist='skew_norm', omega=2, alpha=1.95),
  iso_delay_params=list(dist='Hellewell', shape=1.651524, scale=4.287786),
  sec_infect_params=list(type='Hellewell', disp=0.16),
  import_params=list(type="None"), 
  phys_dist_params=list(
    pd_pop_frac = 0, 
    pd_contact_rate1 = 1.0,
    pd_contact_rate2 = 1.0,
    pd_change_t = 0)
)
start_time <- 0
n_initial <- 20
sim_status <- epi.branch.sim::initialize_sim_status(start_time,n_initial) 
state_df   <- epi.branch.sim::create_state_df(n_initial,sim_params, sim_status, initialize=TRUE)
record_df  <- epi.branch.sim::create_record_df(state_df, sim_status, initialize=TRUE)

# Run for 30 steps
for (t in 1:30){
  out <- epi.branch.sim::step_simulation(sim_status, state_df, record_df, sim_params)
  sim_status <- out$status # update sim_status
  state_df <- out$state    # update state_df
  record_df <- out$record  # update record_df
}

# Output products are the state_df and record_df data frames
```

Detailed documentation on the simulation algorithm, input parameters,
output data frames and examples are provided as vignettes. Once the
package is installed with the `build_vignette=TRUE` option as above,
these vignettes can be viewed with:

``` r
browseVignettes('epi.branch.sim')
```

For help on all documented functions and objects, use:

``` r
help(package='epi.branch.sim')
```

### Project Status

This package is maturing. This is a public release of a tool regularly
used within our team in order to share the tool more widely. At release,
the package is functional, all of the intended features are working and
documented. More features may be added based on usage and demand.
Additional documentation and advanced use examples are also forthcoming.

Plans to improve speed and efficiency are also in the works.

### Getting Help or Reporting an Issue

To report bugs/issues/feature requests, please file an
[issue](https://github.com/bcgov/epi.branch.sim/issues/).

### How to Contribute

If you would like to contribute to the package, please see our
[CONTRIBUTING](CONTRIBUTING.md) guidelines.

Please note that this project is released with a [Contributor Code of
Conduct](CODE_OF_CONDUCT.md). By participating in this project you agree
to abide by its terms.

### License

    Copyright 2020 Province of British Columbia

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software 
    distributed under the License is distributed on an "AS IS" BASIS, 
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and 
    limitations under the License.

------------------------------------------------------------------------

*This project was created using the
[bcgovr](https://github.com/bcgov/bcgovr) package.*


NEWS.md
# epi.branch.sim v1.0.0 (6 August 2020)

* First public GitHub release


R/step_simulation.R
# Copyright 2020 Province of British Columbia
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and limitations under the License.

#' Takes the simulation one dt timestep forward
#'
#' Use the outputs of this function as input for the next step.
#'
#' @param sim_status  \code{sim_status} list containing simulation state vector
#' @param state_df    \code{state_df} object for the simulation
#' @param rec_df      \code{record_df} object for the simulation
#' @param sim_params  \code{sim_params} object (a list) containing simulation parameters
#' @return A list containing
#' \itemize{
#'   \item \code{status}   - The updated \code{sim_status} object
#'   \item \code{state}    - The updated \code{state_df} object
#'   \item \code{record}   - The updated \code{record_df} object
#'   \item \code{new_sec_cases} - Number of new secondary infections generated in this time step
#'   \item \code{new_imp_cases} - Number of new imported infections generated in this time step
#' }
#' @export
step_simulation <- function(sim_status, state_df, rec_df, sim_params){
  # Increment time
  sim_status$t <- sim_status$t + sim_params$dt
  # Update timestamp of current infections
  state_df$days_infected <- state_df$days_infected + sim_params$dt

  # Determine which cases will cause infections this step
  gen_sec_infect_out <- generate_secondary_infections(state_df, sim_params)
  state_df <- gen_sec_infect_out$updated_state_df
  sec_infs_source <- gen_sec_infect_out$sec_infection_sources
  n_sec_infections <- length(sec_infs_source)
  # If new infections, create a temporary state and rec dataframes for the new cases
  # Also update rec_df for the origin cases and sim_status for case_id info
  if (n_sec_infections > 0){
    # Create new state_df for new cases
    sec_cases_state_df <- create_state_df(n_sec_infections,
                                          sim_params,
                                          sim_status,
                                          primary_state_df = state_df,
                                          primary_case_ids = sec_infs_source)
    # Update last_case_id
    sim_status$last_case_id <- max(sec_cases_state_df$case_id)
    # Create new rec_df for new cases
    sec_cases_rec_df <- create_record_df(sec_cases_state_df, sim_status, infection_source=sec_infs_source)
    # TODO: Update rec_df of the origin cases (so no tracking of this yet)
  }

  # Determine whether there will be any imported cases
  # Only import cases at whole number timesteps
  if ( (sim_status$t%%1) < 1e-3 || 1-(sim_status$t%%1) < 1e-3 ){
    gen_imp_infect_out <- generate_imported_infections(sim_params, sim_status)
    n_imp_infections <- gen_imp_infect_out$n_imported_infections
    imp_source <- gen_imp_infect_out$import_source
  } else{
    n_imp_infections<-0
  }
  # If there are imported cases, create temporary state and rec dataframes for the new cases
  # Also update sim_status for case_id info
  if (n_imp_infections > 0){
    # Create new state_df for new imported cases
    imp_cases_state_df <- create_state_df(n_imp_infections,
                                          sim_params,
                                          sim_status,
                                          import=TRUE)
    # Update last_case_id
    sim_status$last_case_id <- max(imp_cases_state_df$case_id)
    # Create new rec_df for new imported cases
    imp_cases_rec_df <- create_record_df(imp_cases_state_df, sim_status, infection_source=imp_source)
  }

  # First, identify and mark all cases where infection has ended, regardless of status
  cases_to_inact <- state_df$days_infected > state_df$infection_length
  n_cases_to_inact <- sum(cases_to_inact)
  if (n_cases_to_inact > 0){
    # Update state_df
    state_df$status[cases_to_inact] <- "inactive"
    # Update rec_df
    ids_to_inact <- state_df$case_id[cases_to_inact]
    rec_df$t_inact[rec_df$case_id %in% ids_to_inact] <- sim_status$t
    rec_df$s_status[rec_df$case_id %in% ids_to_inact] <- "inactive"
  }

  # Identify all remaining cases that are eligible for advancement to next stage
  # Doing it now prevents checking advanced cases for another advancement
  cases_adv_inc <- (state_df$status=="incubation") &
    (
      (state_df$days_infected > state_df$incubation_length) |
      (state_df$days_infected > state_df$incubation_length + state_df$isolation_delay)
    )
  cases_adv_symp <- (state_df$status=="symptomatic") &
    (state_df$days_infected > (state_df$incubation_length + state_df$isolation_delay))
  cases_adv_asymp <- (state_df$status=="asymptomatic") &
    (state_df$days_infected > (state_df$incubation_length + state_df$isolation_delay))

  # Advance current infections from incubation to next stage
  # Goes to symptomatic or asymptomatic based on $is_symptomatic
  # Symptomatic cases that are traced goes right to isolation
  n_cases_adv_inc <- sum(cases_adv_inc)
  if (n_cases_adv_inc > 0){
    # Check whether advancing right to isolation (i.e. delay short enough)
    past_iso_delay <- state_df$days_infected > (state_df$incubation_length + state_df$isolation_delay)
    # Update state_df
    to_symp <- cases_adv_inc & state_df$is_symptomatic & !past_iso_delay
    to_iso  <- cases_adv_inc & past_iso_delay
    to_asymp <- cases_adv_inc & !state_df$is_symptomatic & !past_iso_delay
    state_df$status[to_symp] <- "symptomatic"
    state_df$status[to_iso] <- "isolated"
    state_df$status[to_asymp] <- "asymptomatic"
    # Update rec_df
    # New symptomatic cases
    ids_to_symp <- state_df$case_id[to_symp]
    rec_df$t_symp[rec_df$case_id %in% ids_to_symp] <- sim_status$t
    rec_df$s_status[rec_df$case_id %in% ids_to_symp] <- "symptomatic"
    # New isolated cases
    ids_to_iso <- state_df$case_id[to_iso]
    rec_df$t_symp[rec_df$case_id %in% ids_to_iso] <- sim_status$t
    rec_df$t_iso[rec_df$case_id %in% ids_to_iso] <- sim_status$t
    rec_df$s_status[rec_df$case_id %in% ids_to_iso] <- "isolated"
    # New asymptomatic cases
    ids_to_asymp <- state_df$case_id[to_asymp]
    rec_df$s_status[rec_df$case_id %in% ids_to_asymp] <- "asymptomatic"
  }

  # Advance current infections from symptomatic to isolated after delay
  n_cases_adv_symp <- sum(cases_adv_symp)
  if (n_cases_adv_symp > 0){
    # Update state_df
    state_df$status[cases_adv_symp] <- "isolated"
    # Update rec_df
    ids_to_iso <- state_df$case_id[cases_adv_symp]
    rec_df$t_iso[rec_df$case_id %in% ids_to_iso] <- sim_status$t
    rec_df$s_status[rec_df$case_id %in% ids_to_iso] <- "isolated"
  }

  # Advance current infections from asymptomatic to isolated after delay
  n_cases_adv_asymp <- sum(cases_adv_asymp)
  if (n_cases_adv_asymp > 0){
    # Update state_df
    state_df$status[cases_adv_asymp] <- "isolated"
    # Update rec_df
    ids_to_iso <- state_df$case_id[cases_adv_asymp]
    rec_df$t_iso[rec_df$case_id %in% ids_to_iso] <- sim_status$t
    rec_df$s_status[rec_df$case_id %in% ids_to_iso] <- "isolated"
  }

  # Remove isolated and inactive cases from state_df
  trunc_state_df<-state_df[(state_df$status != 'isolated') & (state_df$status != 'inactive'),]

  # Add secondary infections to state_df and rec_df
  if (n_sec_infections > 0){
    out_state_df <- dplyr::bind_rows(trunc_state_df, sec_cases_state_df)
    out_rec_df <- dplyr::bind_rows(rec_df, sec_cases_rec_df)
  } else{
    out_state_df <- trunc_state_df
    out_rec_df <- rec_df
  }

  # Add imported infections to output state_df and rec_df
  if (n_imp_infections > 0){
    out_state_df <- dplyr::bind_rows(out_state_df, imp_cases_state_df)
    out_rec_df <- dplyr::bind_rows(out_rec_df, imp_cases_rec_df)
  }

  # Return updated inputs
  return(list(status=sim_status, state=out_state_df, record=out_rec_df, new_sec_cases=n_sec_infections, new_imp_cases=n_imp_infections))
}


R/generate_new_infections.R
# Copyright 2020 Province of British Columbia
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and limitations under the License.

# generate_new_infections.R
# Contains functions that tells the model when to add new infections,
# either from secondary infection cases or imported cases.
# New or alternate infection methods should be added here.

#' Adds secondary infections to simulation
#'
#' Potential secondary infections are generated at the time the primary
#' case was generated. This function "loads" the infections into the simulation
#' at the correct time. It also modifies the \code{state_df} to account for the
#' loaded infection.
#'
#' @param state_df    \code{state_df} object for the simulation
#' @param sim_params  \code{sim_params} object (a list) containing simulation parameters
#' @return A list containing
#' \itemize{
#'   \item \code{state_df} - The updated \code{state_df} object for the simulation
#'   \item \code{sec_infection_sources} - A vector of \code{case_id}s causing the infections.
#'   If a case causes multiple secondary infections, the ID is repeated. For example,
#'   c(1,5,5,7) means case 1 caused 1 infection, case 5 caused 2 infections and case 7
#'   caused 1 infection. If no new secondary infections are generated in this time step,
#'   this value is \code{NULL}.
#' }
generate_secondary_infections <- function(state_df, sim_params){
  sec_infection_sources <- c()
  if (nrow(state_df)>0){
    # TODO: Can we replace this for-loop?
    for (row in 1:nrow(state_df)){
      generation_int <- state_df$generation_intervals[[row]]
      # Find which entries in generation_int are causing new secondary infections from this case
      sec_inf_ind <- state_df$days_infected[row] > generation_int
      n_sec_inf <- sum(sec_inf_ind)
      if (n_sec_inf < 1){
        next # no new infections caused by this case this step
      }
      # Add case_id to output list once for each new secondary infection case
      case_id <- state_df$case_id[row]
      sec_infection_sources <- c(sec_infection_sources,rep(case_id,n_sec_inf))
      # Decrease infection counter and remove from generation interval list
      state_df$n_sec_infects[row] <- state_df$n_sec_infects[row] - n_sec_inf
      state_df$generation_intervals[[row]] <- generation_int[!sec_inf_ind]
    }
    return(list(updated_state_df=state_df, sec_infection_sources=sec_infection_sources))
  }
  else{
    return(list(updated_state_df=state_df, sec_infection_sources=NULL))
  }
}

#' Adds imported infections to simulation
#'
#' The number of imported infections may be a deterministic or stochastic value.
#'
#' The list object \code{sim_params$import_params} determines how imported cases are
#' added to the simulation. The following import types are supported, which is chosen by
#' \code{sim_params$import_params$type}:
#' \describe{
#'  \item{\code{type}="None" or "none"}{No cases are imported.}
#'
#'  \item{\code{type}="constant"}{Each day, the number of imported cases is equal to a constant value, set
#'  by \code{sim_params$import_params$rate}.}
#'
#'  \item{\code{type}="constant_two_phase"}{Same as the above, but with two possible constant values. For
#'  times prior to \code{sim_params$import_params$delay}, the rate is \code{sim_params$import_params$rate1}.
#'  After this time, the rate is \code{sim_params$import_params$rate2}.}
#'
#'  \item{\code{type}="daily_constant"}{Each day, the number of imported cases is given by a vector defined by
#'  \code{sim_params$import_params$rate}, which should have length equal to the number of days in the
#'  simulation.}
#'
#'  \item{\code{type}="poisson"}{Each day, the number of imported cases is drawn from a Poisson distribution
#'  with mean set by \code{sim_params$import_params$rate}.}
#'
#'  \item{\code{type}="daily_risk"}{Each day, the number of imported cases is drawn from a Poisson
#'  distribution with mean equal to the daily risk. The daily risk is given by a vector defined by
#'  \code{sim_params$import_params$risk}, which should have length equal to the number of days in the
#'  simulation.}
#' }
#'
#' @param sim_params  \code{sim_params} object (a list) containing simulation parameters
#' @param sim_status  \code{sim_status} object (a list) containing simulation state vector
#' @return A list containing
#' \itemize{
#'   \item \code{n_imported_infections} - Integer with number of imported cases at this time step.
#'   \item \code{import_source} - String describing the source of these new cases. This information
#'   will be recorded in \code{state_df} and \code{record_df}. For now, this will always be "imported".
#' }
generate_imported_infections <- function(sim_params, sim_status){
  import_params <- sim_params$import_params
  if (import_params$type=="None" | import_params$type=='none'){
    n_import<-0
  }
  else if (import_params$type=='constant'){
    n_import <- import_params$rate
  }
  else if (import_params$type=='constant_two_phase'){
    if (sim_status$t < import_params$delay){
      n_import <- import_params$rate1
    } else{
      n_import <- import_params$rate2
    }
  }
  else if (import_params$type=='daily_constant'){
    rate_index <- as.integer(round(sim_status$t))
    n_import <- import_params$rate[rate_index]
  }
  else if (import_params$type=='poisson'){
    n_import <- rpois(1,import_params$rate)
  }
  else if (import_params$type=='daily_risk'){
    risk_index <- as.integer(round(sim_status$t))
    risk<-import_params$risk[risk_index]
    n_import <- rpois(1,risk)
  }
  return(list(n_imported_infections=n_import, import_source='imported'))
}


R/driver.R
# Copyright 2020 Province of British Columbia
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and limitations under the License.

# driver.R
# Contains the functions that will create simulation objects, run a number of simulations
# corresponding to a given scenario, and then tabulate metrics and summary statistics to
# be used in analysis. The functions provided here are meant to be examples and/or
# convienence functions. Users of this package may want to write their own driver
# functions to suit their own analysis needs.

#' Run scenarios as in Hellewell et al. (2020) paper
#'
#' For a given tracing fraction and R0 (a "scenario"), count the fraction of simulations
#' that are controlled. Designed to reproduce results from the source paper for this model: [Hellewell et al. "Feasibility of controlling COVID-19 outbreaks by isolation of cases and contacts." *The Lancet Global Health* 2020; 8: E488-E496. DOI: 10.1016/S2214-109X(20)30074-7](https://doi.org/10.1016/S2214-109X(20)30074-7)
#'
#' The intention of this driver function is to reproduce Figure 3A from Hellewell et al. which
#' shows the fraction of outbreaks controlled as a function of R0 and tracing fraction. These
#' parameters are the positional arguments to this function. The remaining named arguments allow
#' for consideration of other scenarios examined by Hellewell et al. The named arguments' default
#' values are Hellewell et al.'s benchmark values but can be changed to reproduce results from
#' Figure 4 of the source paper.
#'
#' The secondary intention of this driver function is to provide a simple example of how to
#' call the exported functions of this module in order to set up simulations objects, run
#' a number of simulations and return summary statistics/metrics for further analysis.
#'
#' @param nsims     Number of simulations to run in this scenario
#' @param p_trace   Fraction of new infections that are traced
#' @param R0        Basic reproduction number for disease
#' @param infections_before_symptoms  One of "low", "medium" or "high" corresponding to one
#'                                    of three generation interval distributions considered by
#'                                    Hellewell et al. These distributions result in <1%,
#'                                    15%, and 30% of generation interval lengths occuring prior
#'                                    to symptom onset. "Medium" (15%) is the default value
#'                                    and the benchmark used in Hellewell et al.
#' @param iso_delay_length  One of "short" or "long" corresponding to two distributions used
#'                          by Hellewell for the delay from symptom onset to isolation for
#'                          untraced symptomatic cases. "Short" is the default and
#'                          benchmark value.
#' @param p_symp    Fraction of cases that are symptomatic. 1 is the default and benchmark value.
#' @param initial_cases   Number of cases at the start. 20 is the default and benchmark value
#' @param max_cases       Number of total cases before simulation ends with an "epidemic"
#'                        outcome. Defaults to 5000, used by Hellewell et al.
#'
#' @return A named vector with the following entries describing the scenario and
#' the outcome (percentage of simulations that went extinct and absolute number
#' of each type of outcome). Namely the entries are: nsims, R0, p_trace,
#' percentage, nsims_extinct, nsims_epidemic, nsims_ongoing.
#' @export
run_scenario_Hellewell <- function(nsims,p_trace,R0,
                                   infections_before_symptoms='medium',
                                   iso_delay_length='short',
                                   p_symp = 1.0,
                                   initial_cases = 20,
                                   max_cases = 5000){
  col_names <- c("nsims", "R0", "p_trace","percent_controlled", "n_extinct", "n_epidemic", "n_ongoing")
  cat(sprintf('Running %.0f simulations of scenario: R0 = %.2f, p_trace = %.2f\n',nsims,R0,p_trace))
  # Counters for simulation outcomes
  nsims_extinct<-0 # all cases went extinct
  nsims_epidemic<-0  # more than [max_cases] total cases
  nsims_ongoing<-0   # Less than [max_cases] total cases but still new infections after 12 weeks
  for (j in 1:nsims){
    # Set up incubation period length distribution as per Hellewell et al.
    incub_params <- list(dist='weibull', shape=2.322737, scale=6.492272)

    # Set up generation intervals to one of three levels used by Hellewell et al. (called serial intervals there)
    if (infections_before_symptoms == 'low'){ # <1% infections before symptom onset
      generation_int_params <- list(dist='skew_norm', omega=2, alpha=30)
    } else if (infections_before_symptoms == 'medium'){ # 15% infections before symptom onset
      generation_int_params <- list(dist='skew_norm', omega=2, alpha=1.95) # DEFAULT BENCHMARK
    } else if (infections_before_symptoms == 'high'){ # 30% infections before symptoms
      generation_int_params <- list(dist='skew_norm', omega=2, alpha=0.7)
    } else {
      stop('Invalid value for argument "infections_before_symptoms".')
    }

    # Set up isolation delay parameters to one of two levels used by Hellewell et al.
    if (iso_delay_length == 'short'){
      iso_delay_params <- list(dist='Hellewell', shape=1.651524, scale=4.287786) # BENCHMARK
    } else if (iso_delay_length == 'long'){
      iso_delay_params <- list(dist='Hellewell', shape=2.305172, scale=9.483875)
    } else{
      stop('Invalid value for argument "iso_delay_length".')
    }

    # Set up dispersion parameter in number of secondary infections
    sec_infect_params <- list(type='Hellewell', disp=0.16) # From JH2020 "overdispersion in R0"

    # Turn off / disable features not used by Hellewell's work
    infect_dur <- 999 # Hellewell et al. doesn't deactivate cases so we won't either
    do_variable_trace<-FALSE
    p_trace_app<-0
    p_trace_app_comp<-0
    dt <- 1 # Advance one day at a time
    import_params <- list(type='None')
    phys_dist_params <- list(pd_pop_frac=0,pd_contact_rate1=1,pd_contact_rate2=1,pd_change_t=0)


    # Initialize the required simulation objects
    sim_params <- initialize_sim_params(
      R0, infect_dur, do_variable_trace, p_trace,
      p_trace_app, p_trace_app_comp, p_symp, dt,
      incub_params, generation_int_params,
      iso_delay_params, sec_infect_params,
      import_params, phys_dist_params
    )
    sim_status <- initialize_sim_status(start_time=0,initial_cases)
    state_df   <- create_state_df(initial_cases,sim_params,sim_status,initialize=TRUE)
    record_df  <- create_record_df(state_df, sim_status, initialize=TRUE)

    timemax=ceiling(12*7/dt) # 12 weeks after initial cases (Fig 3 caption)
    n_new_cases=rep(0,timemax) # number of new cases per step

    early_exit <- FALSE
    # Run simulation
    for (ii in 1:timemax){
      out <- step_simulation(sim_status, state_df, record_df, sim_params)
      sim_status <- out$status
      state_df <- out$state
      record_df <- out$record
      n_new_cases[ii] <-out$new_sec_cases
      if (nrow(state_df)==0){
        # stop running if no more contagious cases
        nsims_extinct <- nsims_extinct + 1
        early_exit <- TRUE
        break
      } else if(nrow(record_df)>max_cases){
        # also stop running if more than [max_cases] cases (not controlled)
        nsims_epidemic <- nsims_epidemic + 1
        early_exit <- TRUE
        break
      }
    }
    # For runs that finished running, there must still be cases left, so they are "ongoing"
    if (!early_exit){
      if (nrow(state_df)>0){
        nsims_ongoing <- nsims_ongoing+1
      }
    }
  }

  # Create output
  percentage = nsims_extinct/nsims * 100
  results <- c(nsims, R0, p_trace, percentage,nsims_extinct,nsims_epidemic,nsims_ongoing)
  names(results)<- col_names

  return(results)
}

# ABC_start = sim day number to begin ABC matching
# ABC_end = sim day number to end ABC matching
# ABC_match_type = "reported" if using real reports or "adjusted" if adjusting for delay
# ABC_match_last_n = match to last n rows of BC case data file
# ABC_region = "BC" or "VIHA"
run_scenarios <- function(scenario_params, outdir='.',
                          do_ABC=FALSE, ABC_start=7, ABC_end=13,
                          ABC_match_type='reported', ABC_match_last_n=7,
                          ABC_region='BC'){
  scn_id <- scenario_params$scn_id
  nsims <- scenario_params$nsims
  dt <- scenario_params$dt
  tmax_days <- scenario_params$tmax_days
  max_active_cases <- scenario_params$max_active_cases
  save_scn <- scenario_params$save_scn
  import_model <- scenario_params$import_model
  do_variable_trace <- scenario_params$do_variable_trace
  contact_rates_str <- scenario_params$contact_rates
  contact_rates <- eval(parse(text=contact_rates_str))
  pd_p_group_str <- scenario_params$pd_p_group
  pd_p_group <- eval(parse(text=pd_p_group_str))
  pd_delay <- scenario_params$pd_delay
  R0 <- scenario_params$R0
  p_trace <- scenario_params$p_trace
  p_symp <- scenario_params$p_symp
  n_initial <- scenario_params$n_initial

  results=NULL
  for (j in 1:nsims){
    infect_dur <- 999 # Hellewell et al. doesn't deactivate cases so we won't either
    incub_params <- list(dist='lognormal',meanlog=1.57, sdlog=0.65)
    generation_int_params <- list(dist='gamma', shape=2.29, rate=0.36)
    iso_delay_params <- list(dist='uniform_delay', min=2, max=3, delay=2) # 2-3 days if traced, 4-5 if not
    sec_infect_params <- list(type='Hellewell', disp=0.58) # From JH2020 but with larger disp.
    if (import_model=='None'){
      import_params <- 'None'
    }
    else{
      import_params <- get_import_params(file.path('data',paste0(import_model,'.csv')))
    }
    phys_dist_params <- list(contact_rates=contact_rates, p_group=pd_p_group, delay=pd_delay)
    sim_params <- initialize_sim_params(R0, infect_dur, do_variable_trace, p_trace,
                                        p_symp, dt,
                                        incub_params, generation_int_params,
                                        iso_delay_params, sec_infect_params,
                                        import_params, phys_dist_params)
    sim_status <- initialize_sim_status(0,n_initial)
    state_df   <- create_state_df(n_initial, sim_params, sim_status, initialize=TRUE)
    record_df  <- create_record_df(state_df, sim_status, initialize=TRUE)

    timemax=ceiling(tmax_days/dt)
    # early_exit <- FALSE
    # Set up vectors to track key metrics per time step
    n_total=rep(NA,timemax) # number of total cases
    n_active=rep(NA,timemax) # number of active cases
    n_incub=rep(NA,timemax) # number of cases incubating
    n_symp=rep(NA,timemax) # number of cases with symptoms
    n_asymp=rep(NA,timemax) # number of cases not showing symptoms but past incubation
    n_iso=rep(NA,timemax) # number of cases isolated
    n_new_sec=rep(NA,timemax) # number of new secondary cases
    n_new_imp=rep(NA,timemax) # number of new imported cases
    R0_eff=rep(NA,timemax)
    reject_run<-FALSE
    for (ii in 1:timemax){
      # Take a step forward
      out <- step_simulation(sim_status, state_df, record_df, sim_params)
      sim_status <- out$status
      state_df <- out$state
      record_df <- out$record
      # Track key metrics
      n_total[ii] <-nrow(record_df) # all cases ever
      n_active[ii] <-nrow(state_df) # incub + sympt + asympt
      n_incub[ii] <- sum(state_df$status=='incubation')
      n_symp[ii] <- sum(state_df$status=='symptomatic')
      n_asymp[ii] <- sum(state_df$status=='asymptomatic')
      n_iso[ii] <- sum(record_df$s_status=='isolated') # get this from record_df!
      n_new_sec[ii] <-out$new_sec_cases # new sec cases in last dt
      n_new_imp[ii] <- out$new_imp_cases # new imp cases in last dt
      R0_eff[ii] <- mean(record_df$n_sec_infects)
      # If doing ABC, check for match on last day of ABC matching
      if (do_ABC & ii==ABC_end/dt){
        reported_counts <- get_BC_counts(region=ABC_region,last_n=ABC_match_last_n,match_type=ABC_match_type)
        under_report_fac <- scenario_params$under_report_fac # e.g. 4x
        corrected_counts <- reported_counts * under_report_fac
        if (under_report_fac > 1){
          sim_cases <- n_total[seq(ABC_start/dt,ABC_end/dt,1/dt)] # Match with total case counts
        } else{
          sim_cases <- n_iso[seq(ABC_start/dt,ABC_end/dt,1/dt)] # Match with diagnosed cases
        }
        is_under <- any(sim_cases < corrected_counts*0.75) # boolean: any row below 75%?
        is_over <- any(sim_cases > corrected_counts*1.25) # boolean: any row above 125%
        if (is_under | is_over){ # not a match
          reject_run<-TRUE # flag to not save this simulation run into the output df
          break # end sim early
        }
      }
    }
    # Collapse metrics into daily counts (right now assuming dt=1, need a way to sum it up)
    day=0:(timemax*dt)
    bin_every = 1/dt
    nd_total <- c(n_initial,n_total[seq(0,timemax,bin_every)])
    nd_active <- c(n_initial,n_active[seq(0,timemax,bin_every)])
    nd_incub <- c(n_initial,n_incub[seq(0,timemax,bin_every)])
    nd_symp <- c(0,n_symp[seq(0,timemax,bin_every)])
    nd_asymp <- c(0,n_asymp[seq(0,timemax,bin_every)])
    nd_iso <- c(0,n_iso[seq(0,timemax,bin_every)])
    nd_new_S <- c(0,colSums(matrix(n_new_sec,nrow=bin_every)))
    nd_new_I <- c(0,colSums(matrix(n_new_imp,nrow=bin_every)))
    nd_R0eff <- c(R0,R0_eff[seq(0,timemax,bin_every)])
    # Create output dataframe for this run
    results_sim <- data.frame(
      # scenario parameters
      "R0" = R0,
      "p_trace" = p_trace,
      "p_symp" = p_symp,
      "initial_n" = n_initial,
      "sim_id" = sprintf('%.0f.%.0f',scn_id,j), # scenario #, sim #
      # results
      "day" = day,
      "n_total" = nd_total,
      "n_active" = nd_active,
      "n_incub" = nd_incub,
      "n_symp" = nd_symp,
      "n_asymp" = nd_asymp,
      "n_iso" = nd_iso,
      "n_new_S" = nd_new_S,
      "n_new_I" = nd_new_I,
      "R0eff" = nd_R0eff
      )
    # Append to dataframe
    if (!reject_run){
      results <- rbind(results,results_sim)
    }
    # Save a temporary dataframe after each simulation
    if (save_scn){
      s_fname=sprintf('df_scn%04.0f_temp.rds',scn_id)
      saveRDS(results,file=file.path(outdir,'temp',s_fname))
    }
  }
  # Save a final dataframe at the end of the scenario
  if (save_scn){
    s_fname=sprintf('df_scn%04.0f.rds',scn_id)
    saveRDS(results,file=file.path(outdir,s_fname))
  }

  return(results)
}
